### PR TITLE
Add recommendations to Riff Metrics page

### PR DESCRIPTION
**Pull request by jordanreedie**
#### Summary
Adds the 'Recommendations' feature to the Riff metrics page. The goal of these changes is to present the user with predefined 'recommendations' based on how many weeks into the course they are and how they have interacted with their (course|team)mates. 

The course start time is defined as the time that a user with a specific name (currently "timelord") was created. All week numbers, `shouldDisplay` calculations, etc. are based on the existence of this user and assume it was created at the correct time. We use the course start time to calculate which week number the user is currently in. 

The Recommendations themselves (along with the conditions for which they display and the cases in which they are considered 'completed') are static throughout the entire course, so we manually define them all and return all of them in the `getRecommendations` function. They are then filtered by shouldDisplay() and sorted by displayPriority() at the point in time in which they are used. This prevents any issues with time changing after the user opens mattermost (e.g. the user opens a mattermost tab and keeps it open for a week without refreshing). 

The existing tests (located in the `tests` directory) have been disabled, and new tests for the recommendations feature have been added to `riff_tests`. We hope to re-enable the existing tests in the future, after we have returned them to working order.

This also fixes a preexisting bug I encountered during testing where the dashboard would break if the following conditions were met:
1. The user had attempted to join at least one meeting, such that a participant record was created for them in the riff-server database
2. The user had been in no meetings that met the 'usefulness' criteria

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has server changes (please link)
- [ ] Has redux changes (please link)
- [x] Has UI changes
- [ ] Touches critical sections of the codebase (auth, posting, etc.)

### CODE REVIEW

File: actions/post_actions.jsx [hunk 1]
```diff
@@ -255,6 +255,32 @@ export function increasePostVisibility(channelId, focusedPostId) {
...
+export async function getAllPostsFromChannel(channelId) {
+    const currentPosts = await PostActions.getPosts(channelId)(dispatch, getState);
+    if (currentPosts.error) {
+        return [];
+    }
+    const somePost = Object.values(currentPosts.data.posts)[0];
+    const earlyPosts = await PostActions.getPostsBefore(channelId, somePost.id)(dispatch, getState);
+    const latePosts = await PostActions.getPostsAfter(channelId, somePost.id)(dispatch, getState);
+
+    const extractPostsFromQuerry = (querry) => {
```
Comment (jordanreedie) [inline actions/post_actions.jsx]: looks like "query" is misspelled in this function, i'll fix that

File: utils/riff/recommendations/index.jsx [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    );
+
+    const hasConnectedWithCourse = new Recommendation(
+        userId,
+        // eslint-disable-next-line no-magic-numbers
+        5, //first week to display this recommendation
+        3, // number of weeks to display for
+        0, // relative priority to other recs for this week
+        async () => {
+            const userConnections = UserStore.getUserAnalytics()[MMUserAnalyticsDatasets.USER_INTERACTIONS].filter((interaction) => {
```
Comment (jordanreedie) [inline utils/riff/recommendations/index.jsx]: Do we need to change this to use the updated method for getting user analytics?

File: actions/user_actions.jsx [hunk 3]
```diff
@@ -697,3 +697,33 @@ export async function sendPasswordResetEmail(email, success, error) {
...
+ * Send a getUserAnalytics request for a user (userId) and team id (teamId)
+ * Pass a string indicating which dataset to return (name)
+ *
+ * @param {string} userId - The target user for the requested analytics
+ * @param {string} teamId - The target team for the requested analytics
+ * @param {string} name - The name of the type of analytic data to return
+ *
+ * @returns {AsyncThunk<object>} that resolves with the requested analytics
+ */
+export async function getUserAnalytics(userId, teamId, name) {
```
Comment (jordanreedie) [inline actions/user_actions.jsx]: Is this function still being used anywhere, or is this an artifact of the rebase?

File: components/dashboard/components/ChartCard.jsx [hunk 4]
```diff
@@ -52,6 +52,7 @@ const Card = styled.div.attrs({
     ${(props) => props.isNetworkGraphCard === true && `
         height: 100%;
         max-width: 100%;
+        width: 100%;
```
Comment (jordanreedie) [inline components/dashboard/components/ChartCard.jsx]: This addition may be a rebase artifact

File: components/dashboard/components/SectionHeader.jsx [hunk 5]
```diff
@@ -9,7 +9,6 @@ const Title = styled.div`
 font-size: ${(props) => {
         return props.mainHeader ? '40px' : '32px';
     }};
-font-weight: bold;
```
Comment (jordanreedie) [inline components/dashboard/components/SectionHeader.jsx]: rebase artifact?

File: utils/riff/recommendations/index.jsx (hunk 2 above)
Comment (brecriffs) [reply utils/riff/recommendations/index.jsx]: If you connected the component index to the store, you could just pass this info here:
```
const allRecs = getRecommendations(
                this.props.currentUserId,
                this.props.currentTeamId,
                this.props.userLearningGroups,
                startTime,
            );
```

File: actions/user_actions.jsx (hunk 3 above)
Comment (brecriffs) [reply actions/user_actions.jsx]: I think its an ancient artifact.

File: components/dashboard/components/SectionHeader.jsx (hunk 5 above)
Comment (brecriffs) [reply components/dashboard/components/SectionHeader.jsx]: This might be for the the new design for the dashboard.

**Comment by mlippert** [comment]
OK, @jaedoucette @jordanreedie and @brecriffs have given their :+1: to the CR changes.
So I have squashed all of those CR commits into the appropriate commit in this branch prior to merging, and force pushed the new branch.

**Comment by mlippert** [comment]
**Release** [1.0.0-dev.36](https://github.com/rifflearning/mattermost-webapp/releases/tag/1.0.0-dev.36)

**Pull request merged**